src/main/drivers/serial_uart_stm32f30x.c

```diff
@@ -148,44 +124,188 @@ static uartDevice_t uartHardware[] = {
+        .txDMAChannel = UART3_TX_DMA,
+        .pinPair = {
+            { DEFIO_TAG_E(PB11), DEFIO_TAG_E(PB10) },
+            { DEFIO_TAG_E(PC11),  DEFIO_TAG_E(PC10) },
+            { DEFIO_TAG_E(PD9),  DEFIO_TAG_E(PD8) },
+        },
+        .rx = IO_TAG_NONE,
+        .tx = IO_TAG_NONE,
+        .rcc = RCC_APB1(USART3),
+        .af = GPIO_AF_7,
+        .irqn = USART3_IRQn,
+        .txPriority = NVIC_PRIO_SERIALUART3_TXDMA,
+        .rxPriority = NVIC_PRIO_SERIALUART3_RXDMA,
+    },
+
+    // UART4 XXX Not tested (yet!?)
+    // PC11,PC10
+    {
+        .dev = UART4,
+        .rxDMAChannel = UART4_RX_DMA,
+        .txDMAChannel = UART4_TX_DMA,
+        .pinPair = {
+            { DEFIO_TAG_E(PC11), DEFIO_TAG_E(PC10) },
+        },
+        .rx = IO_TAG_NONE,
+        .tx = IO_TAG_NONE,
+        .rcc = RCC_APB1(UART4),
+        .af = GPIO_AF_5,
+        .irqn = UART4_IRQn,
+        .txPriority = NVIC_PRIO_SERIALUART4_TXDMA,
+        .rxPriority = NVIC_PRIO_SERIALUART4_RXDMA,
+    },
+
+    // UART5 XXX Not tested (yet!?)
+    // PD2,PC12
+    {
+        .dev = UART5,
+        .rxDMAChannel = 0,
+        .txDMAChannel = 0,
+        .pinPair = {
+            { DEFIO_TAG_E(PD2), DEFIO_TAG_E(PC12) },
+        },
+        .rx = IO_TAG_NONE,
+        .tx = IO_TAG_NONE,
+        .rcc = RCC_APB1(UART5),
+        .af = GPIO_AF_5,
+        .irqn = UART5_IRQn,
+        .txPriority = NVIC_PRIO_SERIALUART5,
+        .rxPriority = NVIC_PRIO_SERIALUART5,
+    },
 };
 
-static uartDevice_t *uartHardwareMap[5];
+static uartDevice_t *uartHardwareMap[ARRAYLEN(uartHardware)];
+
+#define UART_RX_BUFFER_SIZE 256
+#define UART_TX_BUFFER_SIZE 256
+
+// XXX What a waste...
+// XXX These could be left inside the serialUARTx() function,
+// XXX but they will eventually go away...
+// XXX ARRAYLEN(uartHardware) is overkill; What was the number (count) of UART serial ports???
+
+static volatile uint8_t rxBuffers[UART_RX_BUFFER_SIZE * ARRAYLEN(uartHardware)];
+static volatile uint8_t txBuffers[UART_TX_BUFFER_SIZE * ARRAYLEN(uartHardware)];
 
 void serialInitHardwareMap(serialPinConfig_t *pSerialPinConfig)
 {
-    // Force initialize only for USART1 (PA10, PA9)
-    UNUSED(pSerialPinConfig);
-    uartHardware[0].rx = uartHardware[0].pinPair[0].rx;
-    uartHardware[0].tx = uartHardware[0].pinPair[0].tx;
-    uartHardwareMap[0] = &uartHardware[0];
+    memset(uartHardwareMap, 0, sizeof(uartHardwareMap));
+
+    for (size_t index = 0 ; index < ARRAYLEN(uartHardware) ; index++) {
+        uartDevice_t *uartDev = &uartHardware[index];
+        for (int pair = 0 ; pair < UART_PINPAIR_COUNT ; pair++) {
+
+            if (uartDev->pinPair[pair].rx == pSerialPinConfig->ioTagRx[index]
+                    && uartDev->pinPair[pair].tx == pSerialPinConfig->ioTagTx[index]) {
+                uartDev->rx = uartDev->pinPair[pair].rx;
+                uartDev->tx = uartDev->pinPair[pair].tx;
+                uartDev->rxBuffer = &rxBuffers[index * UART_RX_BUFFER_SIZE];
```

Comment:
It's simple memory reference/address, linker can handle that easily. No loop in this case, but even loop will be fine (with forward declaration)